File: sorter.php
<?php

// sorter quiz
// Copyright (C) 2017 Chris Hendrickson

/** GPLV3 License
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.
  * 
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */

// This code should only run if the script is ran from the commandline
if (!debug_backtrace()) {
	$units = UnitParser::parse("units.txt", " - ");
	foreach($units as $unit) {
		echo $unit->getUnitDisplayString() . PHP_EOL;
	}
}


// Simple data class to encapsulate data
class Unit {
	private $unitNumber;
	private $tenant;

	public function __construct($unitNumber = "", $tenant = "") {
		$this->unitNumber = $unitNumber;
		$this->tenant = $tenant;
	}

	public function getUnitNumber() {
		return $this->unitNumber;
	}

	public function setUnitNumber($value) {
		$this->unitNumber = $value;
	}

	public function getTenantName() {
		return $this->tenant;
	}

	public function setTenantName($value) {
		$this->tenant = $value;
	}

	public function getUnitDisplayString() {
		return $this->unitNumber . " - " . $this->tenant;
	}
}

// Parser class
class UnitParser {

	// Function that returns an array of Units when given a filename and delimiter
	public static function parse($filename, $delimiter) {
		try {
			$file = fopen($filename, 'r');
			$data = [];
			while(!feof($file)) {
				$line = trim(fgets($file));
				
				if($line === "") {
					continue;
				}

				$row = explode($delimiter, $line);
				$unit = trim($row[0], " \t\n\r\0\x0B");
				$tenant = trim($row[1]);
				$data[] = new Unit($unit, $tenant);
			}

			usort($data, array("UnitParser", "sorter"));

			return $data;
		} finally {
			fclose($file);
		}
	}

	// Used to sanitize the apartment number for sorting.
	private static function sanitizeUnitNumber($value) {
		return intval(filter_var($value, FILTER_SANITIZE_NUMBER_INT));
	}

	// Sorter function to compare two unit numbers
	private static function sorter(Unit $a, Unit $b) {
		$aValue = static::sanitizeUnitNumber($a->getUnitNumber());
		$bValue = static::sanitizeUnitNumber($b->getUnitNumber());

		$cmpValue = $aValue - $bValue;

		// If the integer values of the unit numbers are the same, compare the unit numbers as a string.
		if($cmpValue === 0) {
			$cmpValue = strcmp($a->getUnitNumber(), $b->getUnitNumber());
		}

		return $cmpValue;
	}
}

File: sorter.py
#!/usr/bin/env python3

# sorter quiz
# Copyright (C) 2017 Chris Hendrickson

# GPLV3 License
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

class Unit:
  def __init__(self, unit_number: str, tenant: str):
    self.unit_number = unit_number
    self.tenant = tenant

  @property
  def unit_number(self):
    return self._unit_number

  @unit_number.setter
  def unit_number(self, value: str):
    self._unit_number = value

  @property
  def tenant(self):
    return self._tenant

  @tenant.setter
  def tenant(self, value: str):
    self._tenant = value

  @property
  def display_string(self):
    return "%s - %s" % (self.unit_number, self.tenant)

  def __lt__(self, other):
    a_value = Unit._sanitize_unit_number(self.unit_number)
    b_value = Unit._sanitize_unit_number(other.unit_number)

    compare_value = a_value - b_value;

    if compare_value == 0:
      if self.unit_number < other.unit_number:
        compare_value = -1
      elif self.unit_number > other.unit_number:
        compare_value = 1
      else:
        compare_value = 0

    return compare_value < 0

  @staticmethod
  def _sanitize_unit_number(value: str):
    return int(''.join(x for x in value if x.isdigit()))

class UnitParser:
  @staticmethod
  def parse(filename: str, delimeter: str):
    data = []
    with open(filename, 'r') as file:
      for line in file:
        line = line.strip()
         
        if line == "":
          continue

        row = line.split(delimeter)
        unit = row[0].strip()
        tenant = row[1].strip()
        data.append(Unit(unit, tenant))

    return sorted(data)

def main():
  for unit in UnitParser.parse("units.txt", " - "):
    print(unit.display_string);

if __name__ == '__main__':
  main()